frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -3956,6 +4011,145 @@ bool isTypeDefaultInitializable(Context* context, const Type* t) {
+      }
+    } else {
+      auto ast = parsing::idToAst(context, ct->id());
+      const AttributeGroup* attrs = nullptr;
+      if (ast) attrs = ast->attributeGroup();
+      if (checkCopyable && attrs &&
+          (attrs->hasPragma(PRAGMA_SYNC) || attrs->hasPragma(PRAGMA_SINGLE))) {
+        // Syncs and singles are copyable
+        // This is a special case to preserve deprecated behavior before
+        // sync/single implicit reads are removed. 12/8/23
+        fromConst = true;
+      } else {
+        // In general, try to resolve the type's 'init='/'=', and examine it to
+        // determine copy/assignability, respectively.
+        const TypedFnSignature* testResolvedSig =
+            (checkCopyable ? tryResolveInitEq(context, ast, ct, ct)
+                           : tryResolveEq(context, ast, ct));
+        if (testResolvedSig) {
+          if (testResolvedSig->untyped()->isCompilerGenerated()) {
+            // Check for owned class fields.
+
+            bool containsNonNilableOwned = false;
+            bool containsNilableOwned = false;
+            auto resolvedFields =
+                fieldsForTypeDecl(context, ct, DefaultsPolicy::USE_DEFAULTS);
+            for (int i = 0; i < resolvedFields.numFields(); i++) {
+              auto fieldType = resolvedFields.fieldType(i).type();
+              if (auto classTy = fieldType->toClassType()) {
+                if (classTy->decorator().isManaged() &&
+                    classTy->manager()->isAnyOwnedType()) {
+                  if (classTy->decorator().isNonNilable()) {
+                    containsNonNilableOwned = true;
+                  } else {
+                    containsNilableOwned = true;
+                  }
+                }
+              }
+            }
+
+            fromRef = !containsNonNilableOwned && containsNilableOwned;
+            fromConst = !containsNonNilableOwned && !containsNilableOwned;
+          } else {
+            // Check intent of formal to copy/assign from.
+
+            // For init=, formals are (this, other). For =, formals are (lhs,
+            // rhs), unless it is a method in which case they are (this, lhs,
+            // rhs). Get the index of the 'other' or 'rhs' formal.
+            int otherFormalNum = 1;
+            if (!checkCopyable && testResolvedSig->untyped()->isMethod()) {
+              otherFormalNum = 2;
+            }
+            CHPL_ASSERT(testResolvedSig->numFormals() == (otherFormalNum + 1) &&
+                        "unexpected formals");
+            auto otherTy = testResolvedSig->formalType(otherFormalNum);
+            CHPL_ASSERT(!otherTy.isNonConcreteIntent() &&
+                        "should have resolved concrete intent by now");
+            auto otherIntent = otherTy.kind();
+
+            fromConst = (otherIntent == QualifiedType::IN ||
+                         otherIntent == QualifiedType::CONST_IN ||
+                         otherIntent == QualifiedType::CONST_REF);
```

Comment:
IMO it would be reasonable to issue an error here if the `otherIntent` were anything else other than `ref`. E.g. `out` or `inout` are not acceptable for the RHS of `=` / `init=`. I suppose `param` is acceptable and perhaps `type` could be legal as well. I think this is unlikely to come up in practice so a simple error message would do nicely.

Perhaps the nicest way to write it, instead of considering all of the kinds, would be to use `fromConst = uast::isConstQualifier(otherIntent);`.

Reply:
I like the error idea.
For considering the different kinds, `isConstQualifier` wouldn't cover the `IN` intent, so I think I still would have to list them.

Reply:
Or maybe even `isConstQualifier` + `isInQualifier` would be nice, and more future proof.

Reply:
You could also add some kind of sibling to `isConstQualifier` that computes what you want, provided you can give it a good name.

Reply:
I couldn't think of a nice name and couldn't find other places in the code that check this combination of qualifiers so I'll just check them both.